src/d/a/obj/d_a_obj_fairy.cpp

```diff
@@ -1,128 +1,638 @@
 
-SPECIAL_ACTOR_PROFILE(OBJ_FAIRY, dAcObjFairy_c, fProfile::OBJ_FAIRY, 0x166, 0, 2);
+SPECIAL_ACTOR_PROFILE(OBJ_FAIRY, dAcOFairy_c, fProfile::OBJ_FAIRY, 0x22D, 0, 0x23);
 
-STATE_DEFINE(dAcObjFairy_c, Wait);
-STATE_DEFINE(dAcObjFairy_c, Avoid);
-STATE_DEFINE(dAcObjFairy_c, PlayerAvoid);
-STATE_DEFINE(dAcObjFairy_c, CureStart);
-STATE_DEFINE(dAcObjFairy_c, Cure);
-STATE_DEFINE(dAcObjFairy_c, CureEnd);
-STATE_DEFINE(dAcObjFairy_c, CatchDemo);
+static dCcD_SrcSph sSphSrc1 = {
+    /* mObjInf */
+    {/* mObjAt */ {0, 0, {0, 0, 0}, 0, 0, 0, 0, 0, 0},
+     /* mObjTg */ {AT_TYPE_BUGNET | AT_TYPE_BELLOWS | AT_TYPE_WIND, 0x200311, {0, 0, 0x40F}, 0, 0},
+     /* mObjCo */ {0}},
+    /* mSphInf */
+    {30.f},
+};
 
-bool dAcObjFairy_c::createHeap() {
+static dCcD_SrcSph sSphSrc2 = {
+    /* mObjInf */
+    {/* mObjAt */ {0, 0x0, {0, 0, 0}, 0, 0, 0, 0, 0, 0},
+     /* mObjTg */ {0, 0x0, {0, 0, 0x40F}, 0, 0},
+     /* mObjCo */ {0x29}},
+    /* mSphInf */
+    {10.f},
+};
+
+static InteractionTargetDef sAttnTarget = {1, 0, 0, UNK_18, 0x2, 400.0f, 0.0f, 0.0f, -400.0f, 400.0f, 50.0f, 1.0f};
+
+STATE_DEFINE(dAcOFairy_c, Wait);
+STATE_DEFINE(dAcOFairy_c, Avoid);
+STATE_DEFINE(dAcOFairy_c, PlayerAvoid);
+STATE_DEFINE(dAcOFairy_c, CureStart);
+STATE_DEFINE(dAcOFairy_c, Cure);
+STATE_DEFINE(dAcOFairy_c, CureEnd);
+STATE_DEFINE(dAcOFairy_c, CatchDemo);
+
+static void vecCylCalc(mVec3_c &target, const mAng &rot, f32 factor);
+
+bool dAcOFairy_c::createHeap() {
     void *file = getOarcResFile("PutFairy");
     TRY_CREATE(mModel.create(file, "PutFairy", "PutFairy_fly", &mAllocator, 0x120));
 
     return true;
 }
 
-int dAcObjFairy_c::create() {
-    CREATE_ALLOCATOR(dAcObjFairy_c);
+int dAcOFairy_c::create() {
+    CREATE_ALLOCATOR(dAcOFairy_c);
+
+    mStts.SetRank(1);
+    mCcSph1.Set(sSphSrc1);
+    mCcSph1.SetStts(mStts);
+    mCcSph2.Set(sSphSrc2);
+    mCcSph2.SetStts(mStts);
+
+    mObjAcch.Set(this, 1, &mAcchCir);
+    mAcchCir.SetWall(20.0f, 150.0f);
+    mObjAcch.SetWaterNone();
+    mObjAcch.SetGndThinCellingOff();
+    mAcceleration = 0.0f;
+    mMaxSpeed = -40.0f;
+
+    setPosYWaveParams();
+    int rnd = cM::rndInt(40);
+    s16 tmp = mPosYWaveSpeed;
+    mPosYWaveTime = tmp * rnd;
+    mOrigPosition = mPosition;
+    mMaxSpeedY = randMaxSpeedY();
+    mSpeed = mMaxSpeedY;
+
+    field_0xB80 = getFromParams(0, 0xF);
+    field_0xB81 = getFromParams(4, 0x3);
+
+    setSpawnPosition(calcInitialSpawnPosition());
+
+    mVec3_c crrPos = mPosition;
+    mObjAcch.CrrPos(*dBgS::GetInstance());
+    if (mObjAcch.ChkWallHit(nullptr)) {
+        mAng tmp1 = mAcchCir.GetWallAngleY() + cM::rndFX(1.0f) * 1820.0f;
+        mAng tmp2 = mAngle.y;
+        mVelocity.rotY(tmp1 - tmp2);
+        mAngle.y = tmp1;
 
-    // TODO
+        mVec3_c v(mVec3_c::Ez * (cM::rndFX(10.0f) + 150.0f));
+        v.rotY(tmp1);
+
+        mOrigPosition += v;
+        setPosition(crrPos);
+        mOldPosition = mPosition;
+    }
+
+    if (field_0xB80 == 2 || field_0xB80 == 3) {
```

Comment:
field_0xB80 I think is spawn type
Looking at Ghidra for uses of OBJ_FAIRY:

dAcItem_c::spawnItem type = 1

fn_80235280: if (Player Action == DIE || Player Action == RECOVER) type = 3 else type = 2

ActorLink::setFlagCausingSwordUpdate (Not First Fairy && From Bug Net) type = 2

dAcHeartf_c::initializeState_Spore type = 0xF

dAcItem_c::executeState_Wait type = 0xF